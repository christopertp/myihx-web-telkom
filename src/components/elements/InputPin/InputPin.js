import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scoped.css';

export default function InputPin(props) {
  const { className, disabled, error, length, onSubmit } = props;
  const customClass = [
    styles.root,
    error && styles.error,
    className,
  ].filter(Boolean).join(' ');
  const rootRef = useRef(null);
  const [values, setValues] = useState(Array(length).fill(''));

  useEffect(() => {
    rootRef.current.childNodes[0].focus();
  }, []);

  const onChange = i => ({ target }) => {
    if (target.value.match(/[^0-9]/)) {
      return;
    }
    const { childNodes } = rootRef.current;
    const newValues = [...values];

    newValues[i] = target.value;
    setValues(newValues);

    if (i < (length - 1)) {
      [...childNodes].slice(i - (length - 1)).forEach(c => c.value = '');
      childNodes[i + 1].focus();
    } else {
      onSubmit(newValues.join(''));
      setValues(Array(length).fill(''));
      rootRef.current.childNodes[0].focus();
    }
  };

  const onFocus = i => ({ target }) => {
    if (target.value) {
      target.value = '';
      setValues([...values.slice(0, i), target.value, ...Array((length - 1) - i).fill('')]);
    }
  };

  const onKeyUp = i => ({ key }) => {
    if (key === 'Backspace' && i > 0) {
      rootRef.current.childNodes[i - 1].focus();
    }
  };

  return (
    <div className={customClass} ref={rootRef}>
      {[...Array.from({ length }).keys()].map(i => {
        const className = values[i] === '' ? styles.empty : '';
        return (
          <input className={className} disabled={disabled} key={i} maxLength={1}
            onChange={onChange(i)} onFocus={onFocus(i)} onKeyUp={onKeyUp(i)} placeholder="â€¢"
            type="password" value={values[i]} />
        );
      })}
      {error && <small>{error}</small>}
    </div>
  );
}

InputPin.defaultProps = {
  className: '',
  disabled: false,
  error: '',
  length: 6,
  onSubmit: () => {},
};

InputPin.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  length: PropTypes.number,
  onSubmit: PropTypes.func,
};
