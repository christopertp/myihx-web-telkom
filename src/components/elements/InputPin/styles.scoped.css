.root {
  display: flex;
  justify-content: space-between;
  margin: auto;
  margin-bottom: 6.125rem;
  position: relative;
  width: 18.375rem;
}

.root > input {
  background: none;
  border: none;
  border-bottom: 0.125rem solid transparent;
  caret-color: transparent;
  color: #333435;
  font-size: 4.375rem;
  font-weight: 400;
  height: 3.75rem;
  line-height: 3rem;
  outline: none;
  text-align: center;
  width: 3.0625rem;
}

.root > input:focus {
  border-bottom: 0.125rem solid var(--primary-color);
}

.root > input::placeholder {
  color: #cccdce;
}

.root.error > small {
  bottom: -1.75rem;
  color: var(--primary-color);
  display: block;
  font-size: 0.75rem;
  text-align: center;
  position: absolute;
  width: 100%;
}
