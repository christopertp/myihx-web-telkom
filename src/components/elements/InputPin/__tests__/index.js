import React, { useState, useRef } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import InputPin from '../InputPin';

const setV = jest.fn();
const focus = jest.fn();

useRef.mockImplementation(() => ({
  current: {
    childNodes: Array(1).fill({ focus }),
  },
}));

useState.mockImplementation(v => [v, setV]);

describe('src/components/elements/InputPin', () => {
  test('render', () => {
    InputPin.defaultProps.onSubmit();
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<InputPin />);
    expect(tree).toMatchSnapshot();
  });

  test('render error, onFocus, onChange, onKeyUp', () => {
    const onSubmit = jest.fn();
    const props = { ...InputPin.defaultProps, error: 'error', onSubmit };

    useRef.mockImplementationOnce(() => ({
      current: {
        childNodes: Array(InputPin.defaultProps.length).fill({ focus }),
      },
    }));

    const result = InputPin(props);
    expect(result.props.children[1].props.children).toBe('error');

    const input = result.props.children[0][0];
    input.props.onFocus({ target: { value: 'tes' } });
    input.props.onFocus({ target: {} });
    expect(setV).toHaveBeenCalledTimes(1);

    const ev1 = { target: { value: 'a' } };
    expect(input.props.onChange(ev1)).toBe(undefined);
    expect(input.props.onKeyUp(ev1)).toBe(undefined);

    const ev2 = { target: { value: '9' } };
    input.props.onChange(ev2);
    expect(setV).toHaveBeenCalledTimes(2);
    expect(focus).toHaveBeenCalled();

    result.props.children[0][5].props.onChange(ev2);
    expect(onSubmit).toHaveBeenCalled();

    const ev3 = { key: 'Backspace' };
    result.props.children[0][3].props.onKeyUp(ev3);
    expect(focus).toHaveBeenCalled();
  });
});
