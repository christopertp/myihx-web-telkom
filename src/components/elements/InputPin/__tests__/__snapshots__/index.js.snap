// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/InputPin render 1`] = `
<div
  className=""
>
  <input
    disabled={false}
    maxLength={1}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="•"
    type="password"
    value=""
  />
  <input
    disabled={false}
    maxLength={1}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="•"
    type="password"
    value=""
  />
  <input
    disabled={false}
    maxLength={1}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="•"
    type="password"
    value=""
  />
  <input
    disabled={false}
    maxLength={1}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="•"
    type="password"
    value=""
  />
  <input
    disabled={false}
    maxLength={1}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="•"
    type="password"
    value=""
  />
  <input
    disabled={false}
    maxLength={1}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="•"
    type="password"
    value=""
  />
  
</div>
`;
